src/web/Application.php

```diff
@@ -638,7 +645,10 @@ private function _processActionRequest(Request $request): ?BaseResponse
             try {
                 Craft::debug("Route requested: '$route'", __METHOD__);
                 $this->requestedRoute = $route;
-                return $this->runAction($route, $_GET);
+                $response = $this->runAction($route, $_GET);
+                return $response === null && $this->getRequest()->getIsOptions()
+                    ? $this->getResponse()
+                    : $response;
```

Comment:
For `OPTIONS` requests that are action requests and return `null`, we need to return the response instead of null, as it has the manipulated response from the filter, otherwise they will result in a 404 from Yii's handling.